include/watchdog.h

```diff
@@ -1,116 +1,206 @@
-	wdt_api_reload reload;
+	int (*start)(struct device *dev);
+	int (*stop)(struct device *dev);
+	int (*configure)(struct device *dev, struct wdt_cfg *cfg);
+	int (*install_timeout)(struct device *dev, struct wdt_timeout_cfg *cfg);
+	int (*reload)(struct device *dev, void *handle);
 };
 
-static inline void wdt_enable(struct device *dev)
+/**
+ * @brief Enable watchdog instance.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 if successful.
+ * @retval -EFAULT if watchdog instance has not been configured yet.
+ * @retval -EBUSY if watchdog instance has been already started.
+ */
+static inline int wdt_start(struct device *dev)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->enable(dev);
+	return api->start(dev);
 }
 
-static inline void wdt_disable(struct device *dev)
+/**
+ * @brief Disable watchdog instance.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 if successful.
+ * @retval -EFAULT if watchdog instance is not enabled.
+ * @retval -EPERM if watchdog can not be stopped directly by application code.
+ */
+static inline int wdt_stop(struct device *dev)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->disable(dev);
+	return api->stop(dev);
 }
 
-static inline void wdt_get_config(struct device *dev,
-				  struct wdt_config *config)
+/**
+ * @brief Configure watchdog instance.
+ *	  This function is used for configuring global watchdog settings that
+ *	  affect all installed timeouts.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param cfg Pointer to watchdog configuration structure.
+ *
+ * @retval 0 if successful.
+ * @retval -ENOTSUP if any of the set options settings is not supported.
+ */
+static inline int wdt_configure(struct device *dev, struct wdt_cfg *cfg)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->get_config(dev, config);
+	return api->configure(dev, cfg);
 }
 
-static inline int wdt_set_config(struct device *dev,
-				 struct wdt_config *config)
+/**
+ * @brief Install new timeout.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @pararm cfg Pointer to timeout configuration structure.
+ *
+ * @retval 0 if success.
+ * @retval -EBUSY if timeout can't be installed while watchdog has already
+ *		  been started.
+ * @retval -ENOMEM if no more timeouts can be installed.
+ * @retval -EINVAL if any of the window timeout value is out of possible range.
+ * @retval -ENOEXEC if WDT_MODE_CALLBACK is set but interrupt function
+ *		    pointer is NULL.
+ * @retval -EFAULT if timeout with given handle has already been installed.
+ * @retval -ENOTSUP if watchdog supports only one timeout value for all timeouts
+ *		    and supplied timeout window differs from windows for alarms
+ *		    installed so far.
+ */
+static inline int wdt_install_timeout(struct device *dev,
+				      struct wdt_timeout_cfg *cfg)
```

Comment:
Similar whine: wdt_timeout_cfg has only two fields.  Forcing the user to fill them out on the stack and pass a pointer isn't actually saving anything (in fact I'd expect code size would be rather larger) vs. just passing them as "min" and "max" arguments to the function right here.  Seems needlessly complicated to me.

Reply:
@andyross `wdt_timeout_cfg` has 4 fields (or 5 I do not know how to count structure inside structure).